### PR TITLE
For #44568: fixed blocking check for SSO by using a different thread.…

### PR DESCRIPTION
**Pull request by patrick-hubert-adsk**
… Also ensured use of https for SSO site

- Non blocking check for SSO support
- Validation of the URL so that https is used for SSO sites, even if the user has typed 'http://' or no protocol.

### CODE REVIEW

File: python/tank/authentication/login_dialog.py [hunk 1]
```diff
@@ -39,6 +39,47 @@
...
+
+    @url_to_test.setter
+    def url_to_test(self, value):
+        self._url_to_test = value
+
+    def run(self):
+        """
+        Runs the thread.
+        """
+        self.sso_enabled = is_sso_enabled_on_site(self.url_to_test)
```
Comment (patrick-hubert-adsk) [inline python/tank/authentication/login_dialog.py]: The only thing the thread does is the SSO check (using the Shotgun Python API).
We only do a single call to the site. No worries if there are multiple concurrent calls.

The UI is updated by a slot call in the main thread once the 'run()' call is completed.

File: python/tank/authentication/login_dialog.py [hunk 2]
```diff
@@ -86,7 +127,7 @@ def __init__(self, is_session_renewal, hostname=None, login=None, fixed_host=Fal
         # typed, but instead wait for a period of inactivity from the user.
         self._url_changed_timer = QtCore.QTimer(self)
         self._url_changed_timer.setSingleShot(True)
-        self._url_changed_timer.timeout.connect(self._update_ui_according_to_sso)
+        self._url_changed_timer.timeout.connect(self._update_ui_according_to_sso_support)
```
Comment (patrick-hubert-adsk) [inline python/tank/authentication/login_dialog.py]: Fixed the name to something more meaningful.

File: python/tank/authentication/login_dialog.py [hunk 3]
```diff
@@ -148,18 +189,28 @@ def __init__(self, is_session_renewal, hostname=None, login=None, fixed_host=Fal
...
-        self._update_ui_according_to_sso()
+        self._query_task = QuerySiteAndUpdateUITask(self)
+        self._query_task.finished.connect(self._toggle_sso)
+        self._update_ui_according_to_sso_support()
+
+        # We want to wait until we know if the site uses SSO or not, to avoid
+        # flickering GUI.
+        self._query_task.wait()
+
+    def _get_site(self):
```
Comment (patrick-hubert-adsk) [inline python/tank/authentication/login_dialog.py]: We were getting the site and cleanup it up so often, I figured a new method would be useful.

File: python/tank/authentication/login_dialog.py [hunk 4]
```diff
@@ -148,18 +189,28 @@ def __init__(self, is_session_renewal, hostname=None, login=None, fixed_host=Fal
...
         Updates the GUI if SSO is supported or not, hiding or showing the username/password fields.
         """
         # Only update the GUI if we were able to initialize the sam2sso module.
         if self._saml2_sso:
-            url_to_test = self.ui.site.text().encode("utf-8").strip()
-            sso_enabled = is_sso_enabled_on_site(url_to_test)
-            if self._use_sso != sso_enabled:
-                self._toggle_sso()
+            self._query_task.url_to_test = self._get_site()
+            self._query_task.start()
```
Comment (patrick-hubert-adsk) [inline python/tank/authentication/login_dialog.py]: This kicks off the check thread.

File: python/tank/authentication/login_dialog.py [hunk 5]
```diff
@@ -148,18 +189,28 @@ def __init__(self, is_session_renewal, hostname=None, login=None, fixed_host=Fal
...
         self.ui.backup_code.editingFinished.connect(self._strip_whitespaces)
 
-        self._update_ui_according_to_sso()
+        self._query_task = QuerySiteAndUpdateUITask(self)
+        self._query_task.finished.connect(self._toggle_sso)
+        self._update_ui_according_to_sso_support()
+
+        # We want to wait until we know if the site uses SSO or not, to avoid
+        # flickering GUI.
+        self._query_task.wait()
```
Comment (jfboismenu) [inline python/tank/authentication/login_dialog.py]: What if this check takes a long time? The user won't be able to hit cancel and quit.

File: python/tank/authentication/login_dialog.py (hunk 3 above)
Comment (jfboismenu) [reply python/tank/authentication/login_dialog.py]: I just did the same thing in my branch

**Comment by patrick-hubert-adsk** [comment]
@jfboismenu Very good question... And it shows that I don't do GUI or Thread work often... I added a timeout and ensuing logging should we wait too long (e.g. 5 seconds).

This situation should occur only when the web site is under heavy load or if there are connectivity issues.

Otherwise, we fail quickly if the site is unreachable... or we return quickly with the appropriate info if the site is up.

**Pull request merged**